batch.py

```diff
@@ -0,0 +1,394 @@
+# check whether it needs modifying
+def remove_sc_outliers(df):
+    """
+    Remove outlier samples, as identified by sc analysis
+    """
+    outliers = ["966_967", "88_88"]
+    df = df[-df.sampleid.isin(outliers)]
+
+    return df
+
+
+# endregion MISCELLANEOUS
+
+config = get_config()
+
+
+@click.command()
+@click.option("--chromosomes")
+@click.option("--genes")
+@click.option("--celltypes")
+@click.option("--expression-file-prefix")
+@click.option("--sample-mapping-file")
+@click.option("--output-path")
+@click.option("--mt-path")
+@click.option("--anno-ht-path")
+def crm_pipeline(
+    # sc_samples: list["str"],
+    chromosomes: list[str],
+    genes: list[str],
+    celltypes: list[str],
+    expression_files_prefix: str,  # 'scrna-seq/grch38_association_files'
+    sample_mapping_file: str,
+    output_path: str,  # 'tob_wgs_rv/pseudobulk_rv_association'
+    mt_path: str = DEFAULT_JOINT_CALL_MT,  # 'mt/v7.mt'
+    anno_ht_path: str = DEFAULT_ANNOTATION_HT,  # 'tob_wgs_vep/104/vep104.3_GRCh38.ht'
+    window_size: int = 50000,
+):
+
+    sb = hb.ServiceBackend(
+        billing_project=config["hail"]["billing_project"],
+        remote_tmpdir=remote_tmpdir(),
+    )
+    batch = hb.Batch("CellRegMap pipeline", backend=sb)
+
+    # extract samples for which we have single-cell (sc) data
+    sample_mapping_file = remove_sc_outliers(sample_mapping_file)
+    sc_samples = sample_mapping_file["OneK1K_ID"].unique()
+
+    # filter to QC-passing, rare, biallelic variants
+    filter_job = batch.new_python_job(name="MT filter job")
+    output_mt_path = output_path("tob_wgs_rv/densified_rv_only.mt")
+    filter_job.call(
+        filter_variants,
+        mt_path=mt_path,
+        samples=sc_samples,
+        output_mt_path=output_mt_path,
+    )
+
+    # grab all relevant genes across all chromosomes
+    # simpler if gene details are condensed to one file
+    gene_dict: dict[str, dict] = {}
+    for chromosome in chromosomes:
+        geneloc_tsv_path = os.path.join(
+            expression_files_prefix,
+            "gene_location_files",
+            f"GRCh38_geneloc_chr{chromosome}.tsv",
+        )
+
+        # concatenating across chromosomes to have a single dict
+        gene_dict.update(make_gene_loc_dict(geneloc_tsv_path))
+
+    # isolate to the genes we are interested in
+    genes_of_interest = genes or list[gene_dict.keys()]
+
+    # for each gene, extract relevant variants (in window + with some annotation)
+    # submit a job for each gene (export genotypes to plink)
+    genotype_jobs = []
+    for gene in genes_of_interest:
+
+        # final path for this gene - generate first (check syntax)
+        plink_file = output_path(f"{output_path}/plink_files/{gene}")
+        gene_dict[gene]["plink"] = plink_file
+
+        # if the plink output exists, do not re-generate it
+        if to_path(plink_file).exists():
```

Comment:
I'd like to check my understanding here - this file path is being generated as a prefix, and `export_plink` will create a few files (fam/bim/etc)?

If that's the case, this prefix will never exist, and you'll need to check for the existence of a specific file as a proxy for 'was the plink data generated'

```suggestion
        if to_path(f'{plink_file}.bim').exists():
```

Reply:
yup good shout